src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+            : base(CreateMechanism(credential, properties, new DefaultRandomByteGenerator()))
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="GssapiAuthenticator"/> class.
+        /// </summary>
+        /// <param name="username">The username.</param>
+        /// <param name="properties">The properties.</param>
+        public MongoAWSAuthenticator(string username, IEnumerable<KeyValuePair<string, string>> properties)
+            : base(CreateMechanism(username, null, properties, new DefaultRandomByteGenerator()))
+        {
+        }
+
+        /// <inheritdoc/>
+        public override string DatabaseName
+        {
+            get { return "$external"; }
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGO AWS authentication may only use the $external source.", "credential");
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator)
+        {
+            var relativeUri = Environment.GetEnvironmentVariable(awsContainerCredentialsRelativeUriKey);
+            var methodsList = new Func<AwsCredentials>[]
+            {
+                () => AcquireCredentialsFromMongoCredentials(username, securePassword, properties),
+                () => AcquireMechanismDataFromEnvironmentVariables(),
+                () => AcquireMechanismDataFromEcsResponse(ecsBaseUri, relativeUri),
+                () => AcquireMechanismDataFromEc2Response(ec2BaseUri)
+            };
+
+            foreach (var method in methodsList)
+            {
+                var creds = method();
+                if (creds != null)
+                {
+                    return new MongoAWSMechanism(creds.Credentials, creds.SessionToken, randomByteGenerator);
+                }
+            }
+
+            throw new ArgumentException("A MONGODB-AWS must have access key ID.");
+        }
+
+        private static AwsCredentials AcquireCredentialsFromMongoCredentials(string username, SecureString securePassword, IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            var sessionToken = ExtractSessionTokenFromMechanismProperties(properties);
+            ValidateCredentials(username, securePassword, sessionToken);
+
+            if (username == null)
```

Comment:
let's move this validation on the one level up to avoid code duplication

Reply:
> facing more inconsistencies with Secure/Insecure strings
what do you mean here? With the suggested code, you will have a Secure string right at the beginning.
I agree that validation can be a part of the creation step, but at the same time, I don't see a problem making validation as a separate step. Especially if we have other benefits in it like removing code duplication.